File: java_src/15.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int target = -nums[i], j = i + 1, k = nums.length - 1;
            while (j < k) {
                if (nums[j] + nums[k] < target) {
                    j++;
                } else if (nums[j] + nums[k] > target) {
                    k--;
                } else {
                    List<Integer> current = new ArrayList<>();
                    current.add(nums[i]);
                    current.add(nums[j]);
                    current.add(nums[k]);
                    result.add(current);
                    while (++j < k && nums[j] == nums[j - 1]) {}
                    while (--k > j && nums[k] == nums[k + 1]) {}
                }
            }
        }
        return result;
    }
}

File: go_src/15.go
func threeSum(nums []int) [][]int {
    sort.Ints(nums)
    var result [][]int
    for i := 0; i < len(nums) - 2; i++ {
        if i > 0 && nums[i] == nums[i - 1] {
            continue
        }
        target := -nums[i]
        j, k := i + 1, len(nums) - 1
        for j < k {
            if sum := nums[j] + nums[k]; sum < target {
                j++
            } else if sum > target {
                k--
            } else {
                result = append(result, []int{nums[i], nums[j], nums[k]})
                for j++; j < k && nums[j] == nums[j-1]; j++ {}
                for k--; k > j && nums[k] == nums[k+1]; k-- {}
            }
        }
    }
    return result
}
